datafusion/core/src/physical_plan/file_format/parquet.rs

```diff
@@ -497,37 +500,67 @@ impl FileOpener for ParquetOpener {
+                        let col_id =
+                            *predicate.need_input_columns_ids().iter().next().unwrap();
+                        let mut selectors = Vec::with_capacity(row_groups.len());
+                        for r in &row_groups {
+                            let rg_offset_indexes = file_offset_indexes.get(*r);
+                            let rg_page_indexes = file_page_indexes.get(*r);
+                            if let (Some(rg_page_indexes), Some(rg_offset_indexes)) =
+                                (rg_page_indexes, rg_offset_indexes)
+                            {
+                                selectors.extend(
+                                    prune_pages_in_one_row_group(
+                                        &groups[*r],
+                                        &predicate,
+                                        rg_offset_indexes.get(col_id),
+                                        rg_page_indexes.get(col_id),
+                                        &file_metrics,
+                                    )
+                                    .map_err(|e| {
+                                        ArrowError::ParquetError(format!(
+                                            "Fail in prune_pages_in_one_row_group: {}",
+                                            e
+                                        ))
+                                    }),
+                                );
+                            } else {
+                                // fallback select all rows
+                                let all_selected = vec![RowSelector::select(
+                                    groups[*r].num_rows() as usize,
+                                )];
+                                selectors.push(all_selected);
+                            }
+                        }
+                        debug!(
+                            "Use filter and page index create RowSelection {:?} from predicate:{:?}",
+                            &selectors, predicate
+                        );
+                        row_selections.push_back(
+                            selectors.into_iter().flatten().collect::<Vec<_>>(),
                         );
                     }
-                    debug!(
-                        "Use filter and page index create RowSelection {:?} ",
-                        &selectors
-                    );
-                    builder = builder.with_row_selection(RowSelection::from(
-                        selectors.into_iter().flatten().collect::<Vec<_>>(),
-                    ));
+                    let acc = row_selections.pop_front().unwrap_or_default();
+                    let final_selection = row_selections
+                        .into_iter()
+                        .fold(acc, |acc, x| intersect_row_selection(acc, x, true));
+                    builder =
+                        builder.with_row_selection(RowSelection::from(final_selection));
```

Comment:
You can write this like the following (though I am not sure it really any more readable)

```suggestion
                    let final_selection = row_selections
                        .into_iter()
                        .reduce(|acc, x| intersect_row_selection(acc, x, true))
                        .unwrap();
                    builder =
                        builder.with_row_selection(final_selection.into());
```

Reply:
Thanks! forget the `reduce` operator.😂